vit-servicing-station-lib/src/server/snapshot_watcher.rs

```diff
@@ -0,0 +1,204 @@
+    },
+    EventKind, RecursiveMode, Watcher,
+};
+use std::{
+    path::PathBuf,
+    time::{Duration, Instant},
+};
+use tracing::{debug, error, span, trace, warn, Instrument, Level};
+
+type RawSnapshot = Vec<(String, String)>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("invalid snapshot format")]
+    Json(#[from] serde_json::Error),
+
+    #[error(transparent)]
+    PoolError(#[from] diesel::r2d2::PoolError),
+
+    #[error(transparent)]
+    DatabaseError(#[from] diesel::result::Error),
+
+    #[error(transparent)]
+    Notify(#[from] notify::Error),
+
+    #[error("snapshot path is not a file with a proper parent directory")]
+    InvalidPath,
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+}
+
+#[tracing::instrument(skip(context))]
+pub async fn async_watch(path: PathBuf, context: SharedContext) -> Result<(), Error> {
+    if path.is_dir() {
+        return Err(Error::InvalidPath);
+    }
+
+    let file_name = path.file_name().ok_or(Error::InvalidPath)?.to_owned();
+
+    let parent = match path.parent() {
+        Some(parent) => {
+            tokio::fs::create_dir_all(parent).await?;
+            parent.to_path_buf()
+        }
+        None => return Err(Error::InvalidPath),
+    };
+
+    let pool = context.read().await.db_connection_pool.clone();
+
+    load_snapshot_table_from_file(path.clone(), pool.clone()).await?;
+
+    let (tx, rx) = tokio::sync::watch::channel(());
+
+    let watcher = {
+        let watcher_callback_span = span!(Level::DEBUG, "filesystem event");
+
+        tokio::task::spawn_blocking(move || {
+            let mut watcher =
+                notify::recommended_watcher(move |res: notify::Result<event::Event>| {
+                    let _guard = watcher_callback_span.enter();
+
+                    let event = match res {
+                        Ok(event) => event,
+                        Err(e) => {
+                            error!(?e);
+                            return;
+                        }
+                    };
+
+                    if !event
+                        .paths
+                        .iter()
+                        .filter_map(|p| p.file_name())
+                        .any(|p| p == file_name)
+                    {
+                        return;
+                    }
+
+                    trace!(?event);
+
+                    match event.kind {
+                        EventKind::Modify(ModifyKind::Metadata(MetadataKind::WriteTime)) => {}
+                        EventKind::Modify(ModifyKind::Metadata(MetadataKind::Any)) => {}
+                        EventKind::Create(CreateKind::File) => {}
+                        EventKind::Remove(RemoveKind::File) => {}
+                        EventKind::Access(AccessKind::Close(AccessMode::Write)) => {}
+                        EventKind::Modify(ModifyKind::Name(RenameMode::To)) => {}
+                        EventKind::Modify(ModifyKind::Name(RenameMode::Both))
```

Comment:
I had that before, but I removed it mostly because I think it can lead to reading partial writes. I mean, ongoing writes.

When I discussed the task with Caslav he said the idea was to `rsync` the file, and on my machine that only triggers `ModifyKind::Name(Rename::To)` and `ModifyKind::Name(Rename::Both)`, so potentially we only need one of those.

The other ones I left mostly just in case.